### PR TITLE
feat(xdc): reject transactions under the minimum gas price on pool admission

### PR DESCRIPTION
**Pull request by ak88**
## Changes

- Add `MinGasPriceFilter`, an XDC `IIncomingTxFilter` that rejects transactions below the chain's gas price floor, so they stay out of the pool and out of gossip instead of being stored and relayed to peers that drop them.
- Resolve that floor from the chainspec via a new optional `MinGasPrice` engine parameter, mirroring the reference client's two `GetMinGasPrice` variants — subnets and gasless subnets included.
- Set `Blocks.MinGasPrice` to `0` in the XDC mainnet and testnet configs — a **separate live bug**, see below.

### The pool floor

The reference enforces the floor on pool admission only: `ErrZeroGasPrice` and `ErrUnderMinGasPrice` in `core/txpool/txpool.go`. It is **not** a consensus rule — block validation never looks at gas price, so a block carrying an underpriced transaction is still valid and nothing outside the pool checks it. Special transactions (the free block-signing and randomize calls) are exempt. Nethermind had no equivalent, so an XDC node accepted, stored and gossiped transactions every Go peer drops.

The floor is chainspec-resolved rather than hardcoded because a subnet can be configured with a different gas price. `ResolveMinGasPrice` mirrors the reference per fork:

| | `XdcChainSpecEngineParameters` (XDPoSChain) | `XdcSubnetChainSpecEngineParameters` |
|---|---|---|
| floor | stated value below `Gas50xBlock`, `50 ×` it from that block | `50 ×` the stated value, always — the subnet fork has no 50x transition |
| gasless | not available, as in the reference | stated `0` disables the check, as `--gasprice 0` → `common.Gasless` does |
| below the default | raised to `DefaultMinGasPrice`, mirroring the reference's clamp on `--miner-gasprice` so the network floor can't be weakened | same |

Unset means the default, so both shipped chainspecs keep today's 12.5 gwei floor and need no new key. The reference's 50x multiplication is worth knowing about: `--gasprice 1000000000` yields a **50 gwei** floor, not 1 gwei.

### Why `Blocks.MinGasPrice: 0`

Found while testing the above, and worse than it. XDC's base fee is a **constant equal to the floor** (`XdcBaseFeeCalculator`), and EIP-1559 is already active on both networks (`eip1559Transition` 98,800,200 mainnet / 71,550,000 testnet, both behind the shipped pivots). `MinGasPriceTxFilter`, in the XDC block-production pipeline, compares the effective **priority fee**:

```
premiumPerGas = min(gasPrice, feeCap - baseFee) = min(12.5 gwei, 12.5 - 12.5) = 0
```

So with the 1 wei default, a Nethermind validator **skips every transaction paying exactly 12.5 gwei** — the reference's minimum, and what its gas price oracle suggests. There is no equivalent floor in the reference to match: its `NewTransactionsByPriceAndNonce` takes no base fee and orders on the raw gas price, so nothing in its block building converts to an effective tip. It reserves the tip for replacement bumps and overflow eviction only.

That is also why this key cannot carry the pool floor: it is a priority-fee comparison, whereas the reference compares the gas price / fee cap. One key cannot be both.

#### What types of changes does your code introduce?

- [x] Bugfix (a non-breaking change that fixes an issue)
- [ ] New feature (a non-breaking change that adds functionality)
- [ ] Breaking change (a change that causes existing functionality not to work as expected)
- [ ] Optimization
- [ ] Refactoring
- [ ] Documentation update
- [ ] Build-related changes
- [ ] Other: _Description_

#### Requires testing

- [x] Yes
- [ ] No

#### If yes, did you write tests?

- [x] Yes
- [ ] No

#### Notes on testing

`MinGasPriceFilterTests` covers boundary values, the two distinct errors, fee-cap-vs-priority-fee for 1559 transactions, both special-transaction targets, contract creation, a zero (gasless) floor, that locals are not exempt, and end-to-end rejection on `TxPool.SubmitTx`. `XdcSpecProviderTests` covers the resolution: the 50x flip landing on its own release-spec boundary, the clamp, unset, zero outside a subnet, and the subnet's always-raised and gasless cases. `BlockProductionMinGasPriceTests` pins the `Blocks.MinGasPrice` behaviour. Full `Nethermind.Xdc.Test` (735) and `Nethermind.Runner.Test` (769) suites pass.

Four existing tests failed once the floor applied and now use a realistic gas price, which models a live node better.

### Verified on live nodes, before and after

A four-validator chain: two Nethermind validators and two `xinfinorg/devnet:latest` Go validators, round-robin proposers, EIP-1559 active with base fee exactly 12.5 gwei. Each case is an independent throwaway sender submitting one transaction.

**Pool admission.** Before, on `nethermindeth/nethermind:master`; after, on an image built from this branch:

| gas price | master | **this branch** | go devnet |
|---|---|---|---|
| 0 | ACCEPTED | **REJECTED** — `zero gas price` | REJECTED — `zero gas price` |
| 1 wei | ACCEPTED | **REJECTED** — `under min gas price` | REJECTED — `under min gas price` |
| 0.25 gwei | ACCEPTED | **REJECTED** — `under min gas price` | REJECTED — `under min gas price` |
| 12.5 gwei - 1 | ACCEPTED | **REJECTED** — `under min gas price` | REJECTED — `under min gas price` |
| 12.5 gwei | ACCEPTED | ACCEPTED | ACCEPTED |
| 13.75 gwei | ACCEPTED | ACCEPTED | ACCEPTED |

Four of six diverged before; all six agree after, with the reference's own error wording. On master, the four that Go refused were **still pending 190 blocks later**, 95 of them Nethermind-proposed — unmineable by any validator on the network, which is the pool pollution this filter removes. The 12.5 gwei floor was resolved from that chain's chainspec (`Gas50xBlock = 0`, `XDPoS` engine → `DefaultMinGasPrice × 50`), so the new resolution path is exercised on a real chain rather than only in unit tests.

**Block production**, isolating the config value on otherwise identical nodes — same image, same chain, same moment, masternode1 with `--Blocks.MinGasPrice=0` and masternode2 with the 1 wei default. Eight senders, one transaction each at exactly the floor, all submitted to Nethermind; all eight mined:

| proposer | took |
|---|---|
| **NM-mn1** (`MinGasPrice=0`) | 2 |
| **NM-mn2** (1 wei default) | **0** |
| GO-mn3 | 6 |

Across that window masternode2 proposed 35 blocks and included 12 transactions — sign transactions and above-floor ones — but not one of the eight at the floor. The config value alone separates the two nodes.

#### Requires documentation update

- [ ] Yes
- [x] No

#### Requires explanation in Release Notes

- [ ] Yes
- [x] No

### One decision for reviewers

Setting `Blocks.MinGasPrice: 0` in the two shipped configs reaches **only operators who use those configs**. The live chain above runs its own config file, and the block-production bug reproduced on it with the branch image — so every subnet deployment and any operator with a custom config still runs a validator that silently skips minimum-price transactions.

Defaulting the key from the XDC plugin instead would reach them, but I did not find an established pattern for a plugin overriding another module's config default, and "did the operator set it explicitly" is not cleanly detectable from `IBlocksConfig`. Flagging it rather than guessing: happy to implement whichever way you prefer.

### Deliberate deviations

Three things about the filter differ from what a reader of the neighbouring XDC filters might expect, all to follow the reference, and all documented on the type: locals are not exempt; the fee cap is compared rather than the priority fee; and the exemption is `IsSpecialTransaction`, narrower than the `RequiresSpecialHandling` set `BuyGas` gives free gas to — the reference keeps XDCX order and lending traffic in separate pools, and both shipped chainspecs disable it outright via `TIPXDCXReceiverDisable`, so it never reaches this filter.

The spec is read at `HeadNumber + 1`, matching `BlackListedAddressFilter` and `SignTransactionFilter`, where the reference reads its current header. The two differ only for the single block at `Gas50xBlock`, which coincides with the XDPoS 2.0 switch block on both networks and is long past.

### Remaining divergences in the pool's gas and priority handling, not in scope

- **`MaxPendingTxsPerSender`** defaults to 0 (unlimited) against the reference's `LimitThresholdNonceInQueue = 10`, and measures from the confirmed rather than the pending nonce. One config line; the cheapest remaining convergence.
- **Pool capacity** — `TxPoolConfig.Size` 2048 against the reference's 5120 slots, and no per-account cap against its `AccountSlots = 16`. Different drop thresholds under load, same eviction semantics.
- **Special-transaction replacement** — the reference refuses to replace a pending special transaction and lets one force-replace a non-special transaction, bypassing the price bump; Nethermind does the opposite in both directions. Only reachable from the validator's own key, and `SignTransactionManager` takes its nonce from `GetLatestPendingNonce` so it queues rather than colliding. Fixing it needs a new injection point in `Nethermind.TxPool` for a custom replacement comparer.
- **`eth_gasPrice`** — the reference clamps its oracle up to `GetMinGasPrice` on non-1559 heads; Nethermind's clamps to `Blocks.MinGasPrice`, now 0 on XDC. On the live chain Nethermind suggests 13,750,000,001 against Go's 12,500,000,000, so it stays above the floor either way, but it's worth its own issue.

No change is needed for the reference's non-local underpriced check: once 1559 is active it compares the fee cap against the base fee, which on XDC is arithmetically the floor this filter already enforces.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs [hunk 1]
```diff
@@ -5,5 +5,9 @@ namespace Nethermind.Xdc.Spec;
 
 public class XdcSubnetChainSpecEngineParameters : XdcChainSpecEngineParameters
 {
+    // The subnet reference client has no 50x transition - its GetMinGasPrice always returns MinGasPrice50x - so the
+    // raised gas price floor applies from genesis. A chainspec that states Gas50xBlock still overrides this.
+    public XdcSubnetChainSpecEngineParameters() => Gas50xBlock = 0;
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs]: Right that the rationale was wrong — moving zero is a no-op, so a zero-price transaction does work on Nethermind today up to a point. Two corrections in the other direction though: it is not end-to-end working (a Nethermind validator already cannot include one, because `Blocks.MinGasPrice` defaults to 1 wei and the priority fee is 0), and there is no `--gasless` flag — `common.Gasless` is derived from `--gasprice 0`, a node-local setting that also zeroes `GetGasFee`, which `core/state_processor.go` reads for TRC21 issuer accounting, so in the reference it is consensus-visible despite being per-node.

Verified your `GetMinGasPrice` question against the XDC-Subnet source: for any non-nil header it returns `MinGasPrice50x` unconditionally, and `MinGasPrice` only for a nil one. So 12.5 gwei is right for subnets.

The subnet parameter is gone entirely now. `Gas50xBlock` coincides with the XDPoS 2.0 switch block on both networks (mainnet 80,370,000, testnet 56,828,700) and only 2.0 is supported, so the 0.25 gwei era is unreachable and the floor is one constant everywhere — no spec property, no transition, no subnet default, so no conflation left to worry about. I have left out the chainspec key for now and said so in the PR: if a gasless subnet actually turns up, that is the point to design the escape hatch rather than guess its shape.

File: src/Nethermind/Nethermind.Xdc/Spec/XdcChainSpecBasedSpecProvider.cs [hunk 2]
```diff
@@ -81,6 +81,9 @@ protected override ReleaseSpec CreateReleaseSpec(ChainSpec chainSpec, ulong rele
         releaseSpec.IsTIPXDCXMiner = (chainSpecEngineParameters.TipXDCX ?? ulong.MaxValue) <= releaseStartBlock && releaseStartBlock < (chainSpecEngineParameters.TIPXDCXMinerDisable ?? ulong.MaxValue);
         releaseSpec.IsTIPXDCXReceiver = (chainSpecEngineParameters.TipXDCX ?? ulong.MaxValue) <= releaseStartBlock && releaseStartBlock < (chainSpecEngineParameters.TIPXDCXReceiverDisable ?? ulong.MaxValue);
         releaseSpec.IsDynamicGasLimitBlock = (chainSpecEngineParameters.DynamicGasLimitBlock ?? ulong.MaxValue) <= releaseStartBlock;
+        releaseSpec.MinimumGasPrice = (chainSpecEngineParameters.Gas50xBlock ?? ulong.MaxValue) <= releaseStartBlock
+            ? XdcConstants.DefaultMinGasPrice * XdcConstants.Gas50xMultiplier
+            : XdcConstants.DefaultMinGasPrice;
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/Spec/XdcChainSpecBasedSpecProvider.cs]: Resolved by deletion — the derivation is gone. `Gas50xBlock` coincides with the XDPoS 2.0 switch block on both networks and only 2.0 is supported, so the floor is a single `XdcConstants` entry with no spec property, transition or subnet default to test.

Chasing this did turn up something the tests were hiding, though, and it is now covered: `BlockProductionMinGasPriceTests`. XDC's base fee is a constant equal to the floor and EIP-1559 is already active on both networks, so `MinGasPriceTxFilter` computes `min(12.5 gwei, 12.5 − 12.5) = 0` for a transaction paying exactly the minimum and, with the 1 wei default, the block producer skips it. Fixed by setting `Blocks.MinGasPrice: 0` in both configs.

File: src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+        if (minimum.IsZero || tx.IsSpecialTransaction(spec))
+            return AcceptTxResult.Accepted;
+
+        // MaxFeePerGas is the fee cap of a 1559 transaction and the gas price of a legacy one, matching what the
+        // reference client compares - its tx.GasPrice() returns GasFeeCap for dynamic fee transactions.
+        if (tx.MaxFeePerGas < minimum)
+        {
+            AcceptTxResult result = AcceptTxResult.FeeTooLow.WithMessage($"Gas price below the chain minimum {tx.MaxFeePerGas} < {minimum}");
+            if (_logger.IsTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, {result}.");
+            return result;
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs]: Both done. The floor is a constant now, so the rejection result is a `static readonly` built once, and `Metrics.PendingTransactionsTooLowFee` is bumped — its description talks about the lowest fee in the pool, but `FeeTooLowFilter` already uses it for its zero-fee branch, so a chain-minimum drop fits the same counter without adding surface.

File: src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+internal sealed class MinGasPriceFilter(
+    IChainHeadInfoProvider chainHeadInfoProvider,
+    ISpecProvider specProvider,
+    ILogManager logManager) : IIncomingTxFilter
+{
+    private readonly ILogger _logger = logManager.GetClassLogger<MinGasPriceFilter>();
+
+    public AcceptTxResult Accept(Transaction tx, ref TxFilteringState state, TxHandlingOptions txHandlingOptions)
+    {
+        IXdcReleaseSpec spec = specProvider.GetXdcSpec(chainHeadInfoProvider.HeadNumber);
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs]: Switched to `HeadNumber + 1` to match the siblings. It is moot now rather than a judgement call: with the floor a constant, the spec is only consulted for the special transaction addresses, so the block it is read for cannot change the outcome. The comment says exactly that.

File: src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        UInt256 minimum = spec.MinimumGasPrice;
+
+        if (minimum.IsZero || tx.IsSpecialTransaction(spec))
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs:60]: Confirmed and noted in the remark: the reference keeps XDCX order and lending traffic in its own pools (`IsXDCXApplyTransaction` / `IsXDCZApplyTransaction` are validated separately in `validateTx`, and the order/lending pools are distinct), so those transactions never reach this filter. `IsSpecialTransaction` is the right set, matching the reference's own guard on the floor check.

File: src/Nethermind/Nethermind.Xdc.Test/MinGasPriceFilterTests.cs [hunk 6]
```diff
@@ -0,0 +1,144 @@
...
+using NSubstitute;
+using NUnit.Framework;
+
+namespace Nethermind.Xdc.Test;
+
+[Parallelizable(ParallelScope.All)]
+internal class MinGasPriceFilterTests
+{
+    private const ulong MinGasPrice = XdcConstants.DefaultMinGasPrice;
+    private static readonly Address BlockSigner = new("0x00000000000000000000000000000000b000089");
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc.Test/MinGasPriceFilterTests.cs]: Fixed — the test now uses `TestItem.AddressA`/`AddressB` for the two special targets rather than a hand-written literal, which also lets it cover the randomize address as well as the block signer.

File: src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs]: Reversed course on this, and you were right to push: a subnet can be configured with a different gas price, so a hardcoded floor cannot describe every XDC chain. There is now an optional `MinGasPrice` engine parameter, and `ResolveMinGasPrice` mirrors the reference client per fork rather than encoding one fork in terms of the other:

- `XdcChainSpecEngineParameters` is XDPoSChain's `GetMinGasPrice` — the stated floor below `Gas50xBlock`, raised 50x from it.
- `XdcSubnetChainSpecEngineParameters` overrides it with the subnet fork's — no 50x transition, always raised, and a stated zero means gasless, which its `--gasprice 0` enables.

So the `Gas50xBlock = 0` conflation you flagged is gone: the subnet override does not consult that block at all. A stated floor below `DefaultMinGasPrice` is raised to it, mirroring how the reference clamps `--miner-gasprice`, and unset means the default, so both shipped chainspecs keep today's 12.5 gwei floor with no new key. The interface doc no longer promises an escape hatch that only tests can reach — a subnet chainspec states `0` and gets one.

Worth adding on the gasless mechanics, since it shaped the design: the reference has no `--gasless` flag. `common.Gasless` is derived from `--gasprice 0`, a node-local setting that also zeroes `GetGasFee`, which `core/state_processor.go` reads for TRC21 issuer accounting — so in the reference it is consensus-visible despite being per-node. Encoding it in the chainspec instead makes it a chain-level statement, which is the safer half of that trade.

File: src/Nethermind/Nethermind.Xdc/Spec/XdcChainSpecBasedSpecProvider.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/Spec/XdcChainSpecBasedSpecProvider.cs]: Now covered, and the derivation is back — see the reply on the subnet thread for why. `XdcSpecProviderTests` follows the `DynamicGasLimitBlock_takes_effect_on_its_own_block` pattern you pointed at:

- `MinimumGasPrice_is_raised_50x_on_Gas50xBlock` — the flip lands on its own release-spec boundary, asserting the default on both sides of it.
- `MinimumGasPrice_follows_the_stated_floor` — unset, below the default (clamped up), zero outside a subnet (not gasless), and above the default, each checked before and after the 50x block.
- `MinimumGasPrice_on_a_subnet_is_raised_from_genesis_and_zero_means_gasless` — the subnet override at block 0 and at 1,000,000, plus the gasless case.

Eight new cases, 735 in the suite, all green.

File: src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs]: Both done, and the first one landed better than a cached result would have. Mirroring the reference means reporting its two distinct errors, `zero gas price` and `under min gas price`, and its own errors carry no numbers — so these are two `static readonly` results with no interpolation on any path, and the floor is logged at trace level instead. `Metrics.PendingTransactionsTooLowFee` is bumped on rejection: its description talks about the lowest fee in the pool, but `FeeTooLowFilter` already uses it for its zero-fee branch, so a chain-minimum drop fits the same counter without adding surface.

File: src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/TxPool/MinGasPriceFilter.cs]: Correcting my earlier reply on this thread — it is stale, and in a way that matters for your point.

I said the block no longer mattered because "with the floor a constant, the spec is only consulted for the special transaction addresses". That was true of the intermediate revision, but the floor is now resolved from the chainspec again, so `spec.MinimumGasPrice` is read at `HeadNumber + 1` and the block genuinely does determine the value at the `Gas50xBlock` boundary. Your original observation therefore still stands rather than being moot.

The current position, deliberately: `HeadNumber + 1` matches `BlackListedAddressFilter` and `SignTransactionFilter`, i.e. the spec of the block the transaction would land in, and it diverges from the reference client — which reads `pool.chain.CurrentHeader().Number` — for exactly the one block at `Gas50xBlock`. I picked sibling consistency over that one-block parity because an admission filter deciding against the block a transaction can actually appear in seems the more defensible rule, and because on both shipped networks `Gas50xBlock` coincides with the XDPoS 2.0 switch block and is long past, so the boundary is unreachable in practice.

Happy to flip it to `HeadNumber` for exact parity if you would rather the reference win here; it is a one-word change and the tests pin the behaviour either way.

**Review by benaadams — APPROVED** [review]

File: src/Nethermind/Nethermind.Xdc.Test/BlockProductionMinGasPriceTests.cs [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+/// nothing in its block building converts to an effective tip.
+/// </remarks>
+[Parallelizable(ParallelScope.All)]
+internal class BlockProductionMinGasPriceTests
+{
+    [TestCase(0ul, true, TestName = "Zero floor includes a transaction paying the chain minimum")]
+    [TestCase(1ul, false, TestName = "Default one wei floor skips a transaction paying the chain minimum")]
+    public void MinGasPriceTxFilter_TxAtChainMinimum_WithXdcBaseFee(ulong configuredFloor, bool expectedAllowed)
+    {
+        XdcReleaseSpec spec = new() { IsEip1559Enabled = true, BaseFeeCalculator = new XdcBaseFeeCalculator() };
```
Comment (benaadams) [inline src/Nethermind/Nethermind.Xdc.Test/BlockProductionMinGasPriceTests.cs:32]: [P3] Test the shipped configuration fix

This test constructs BlocksConfig with the desired value, so reverting both JSON changes would leave it passing. Please add coverage that loads the shipped xdc.json and xdc-testnet.json configurations and verifies that their production filters accept a transaction priced exactly at the base fee, so the configuration fix itself has a regression test.

**Comment by benaadams** [comment]
@claude review

File: src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs [hunk 8]
```diff
@@ -1,9 +1,23 @@
...
+    /// <remarks>
+    /// The subnet fork's <c>common.GetMinGasPrice</c> has no 50x transition - it always returns the raised floor - and
+    /// honours a gasless node, which its <c>--gasprice 0</c> enables. A subnet chainspec therefore states
+    /// <see cref="XdcChainSpecEngineParameters.MinGasPrice"/> as zero to run gasless, and
+    /// <see cref="Gas50xBlock"/> is not consulted at all.
+    /// </remarks>
+    internal override UInt256 ResolveMinGasPrice(ulong blockNumber) =>
+        MinGasPrice == UInt256.Zero
+            ? UInt256.Zero
+            : ConfiguredMinGasPrice * XdcConstants.Gas50xMultiplier;
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs:29]: Documented, and the constraint is firmer than either of us put it: the subnet fork's `Header` struct has **no base fee field at all**. XDPoSChain's carries `BaseFee *big.Int \`rlp:"optional"\`` with the EIP-1559 comment; XDC-Subnet's ends at `Penalties []common.Address` with nothing after it. Nethermind mirrors this — `XdcHeaderDecoder` encodes `BaseFeePerGas` as an optional tail when non-zero, and `XdcSubnetHeaderDecoder` has no base fee handling whatsoever.

So a subnet chainspec that set `eip1559Transition` would not merely get incoherent fee rules: the value could not round-trip through the encoder, so the header hash would not match anything peers sent or expected. The configuration is unreachable rather than merely unwise, which is why I took the documentation route you suggested over deriving the base fee from the resolved floor — deriving would add machinery for a case the encoder already rejects, and it would widen the blast radius to every XDC chain's base fee.

The `<remarks>` now states exactly that. One refinement to your failure analysis, for the record: with 1559 on, `FeeTooLowFilter` rejects a *gossiped* zero-price transaction (its zero check is gated on `isEip1559Enabled`, and `CalculateGasPrice` returns `tx.GasPrice` for a legacy transaction), so the shape would be "accepted locally via RPC, rejected by every peer, never mined" rather than pooled everywhere. Same end state, narrower path.

File: src/Nethermind/Nethermind.Xdc.Test/Helpers/XdcTestBlockchain.cs [hunk 9]
```diff
@@ -604,7 +605,9 @@ public PrivateKey[] TakeRandomMasterNodes(IXdcReleaseSpec headSpec, EpochSwitchI
 
     public TransactionBuilder<Transaction> CreateTransactionBuilder()
     {
-        TransactionBuilder<Transaction> txBuilder = BuildSimpleTransaction;
+        // MinGasPriceFilter keeps anything cheaper out of the pool, so a transaction below the floor would never
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc.Test/Helpers/XdcTestBlockchain.cs]: Correct, and this was a regression I introduced without noticing. Those `.WithGasPrice(...)` lines *were* load-bearing when the floor was a hardcoded constant — four tests failed and that is why I added them. Making the floor chainspec-resolved quietly demoted them to decoration, and left the comment asserting something untrue of the harness.

Taken the better option: `WrapReleaseSpec` now sets `MinimumGasPrice`, so the filter is live across the whole XDC suite and those gas prices are fixtures the tests depend on again. The proof is that `SubmitTx_UnderMinGasPrice_IsRejectedOnPoolAdmission` no longer calls `ChangeReleaseSpec` at all — it relied on that override before, and now relies solely on the harness floor. 737 green.

File: src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs [hunk 10]
```diff
@@ -1,9 +1,23 @@
...
+
+    /// <inheritdoc/>
+    /// <remarks>
+    /// The subnet fork's <c>common.GetMinGasPrice</c> has no 50x transition - it always returns the raised floor - and
+    /// honours a gasless node, which its <c>--gasprice 0</c> enables. A subnet chainspec therefore states
+    /// <see cref="XdcChainSpecEngineParameters.MinGasPrice"/> as zero to run gasless, and
+    /// <see cref="Gas50xBlock"/> is not consulted at all.
+    /// </remarks>
+    internal override UInt256 ResolveMinGasPrice(ulong blockNumber) =>
+        MinGasPrice == UInt256.Zero
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/Spec/XdcSubnetChainSpecEngineParameters.cs]: Applied, with a one-line comment saying why it matters, since unset-vs-explicit-zero is the whole semantics of the override:

```csharp
// Unset is not zero: only an explicit zero means gasless.
MinGasPrice is { IsZero: true }
```

File: src/Nethermind/Nethermind.Xdc/Spec/XdcChainSpecEngineParameters.cs [hunk 11]
```diff
@@ -76,6 +88,18 @@ public List<V2ConfigParams> V2Configs
...
+    /// <remarks>Mirrors <c>common.GetMinGasPrice</c> of XDPoSChain: the configured floor below
+    /// <see cref="Gas50xBlock"/>, raised 50x from it.</remarks>
+    internal virtual UInt256 ResolveMinGasPrice(ulong blockNumber) =>
+        (Gas50xBlock ?? ulong.MaxValue) <= blockNumber
+            ? ConfiguredMinGasPrice * XdcConstants.Gas50xMultiplier
+            : ConfiguredMinGasPrice;
+
+    /// <summary>The stated floor, never below <see cref="XdcConstants.DefaultMinGasPrice"/>.</summary>
+    protected UInt256 ConfiguredMinGasPrice =>
+        UInt256.Max(MinGasPrice.GetValueOrDefault(), XdcConstants.DefaultMinGasPrice);
```
[the remark this answers is not in the document]
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc/Spec/XdcChainSpecEngineParameters.cs:101]: Applied — `private protected`, matching `ResolveMinGasPrice`'s `internal virtual`.

File: src/Nethermind/Nethermind.Xdc.Test/BlockProductionMinGasPriceTests.cs (hunk 7 above)
Comment (ak88) [reply src/Nethermind/Nethermind.Xdc.Test/BlockProductionMinGasPriceTests.cs:32]: Good catch — you were right that reverting both JSON edits left everything green.

Added `Xdc_produces_blocks_with_transactions_priced_at_the_base_fee` to `ConfigFilesTests`, one case per shipped config. Rather than construct a `BlocksConfig`, it loads the shipped `xdc.json` / `xdc-testnet.json` *and* resolves the base fee through the real chainspec — `ChainSpecFileLoader` → `XdcChainSpecBasedSpecProvider` → the spec at `Eip1559Transition` → `BaseFeeCalculator.Calculate` — then asserts `MinGasPriceTxFilter` admits a legacy transaction priced exactly at it. So it pins the chainspec and the config together, not just the config value.

Verified it actually bites: with `"MinGasPrice": 0` removed from `xdc.json`, the mainnet case fails with

```
transaction underpriced, EffectivePriorityFeePerGas too low 0 < 1, BaseFee: 12500000000
```

while the untouched `xdc-testnet.json` case stays green. Restored, both pass, and it sits next to your existing `Xdc_archive_syncs_from_the_XDPoS_v2_switch_block`.

Worth flagging one thing this exposed, since it explains why nothing caught the bug earlier: `TestBlockchain.CreateConfigs()` hardcodes `MinGasPrice = 0` ("Tx pool test seems to need this"), so the entire integration harness has been masking it.

That leaves one open question in the PR description I would value your view on: setting the key in the two shipped configs only reaches operators who use them. On a live four-validator chain running its own config file, the block-production bug still reproduced with this branch's image. Defaulting it from the XDC plugin would reach everyone, but I did not find an established pattern for a plugin overriding another module's config default, and explicit-vs-default is not cleanly detectable from `IBlocksConfig`.

**Review by batrr — APPROVED** [review]
